File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    boolean[][] visited;
    int[] xMove = {1,0,-1,0};
    int[] yMove = {0,1,0,-1};
    
    public int bfs(int x, int y, int n, int m, int[][] maps){
        visited[x][y] = true;
        Queue<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{0,0,0});
        while(!q.isEmpty()){
            int[] tmp = q.poll();
            int tmpX = tmp[0];
            int tmpY = tmp[1];
            int tmpDepth = tmp[2];
            
            if (tmpX == n-1 && tmpY == m-1) {
                return tmpDepth+1;
            }
            
            for(int i=0; i<4; i++){
                int newX = tmpX+xMove[i];
                int newY = tmpY+yMove[i];
                if(newX<0 || newY<0 || newX>=n || newY>=m) continue;
                
                if(visited[newX][newY] || maps[newX][newY]==0) continue;
                    int[] newTmp = {tmpX+xMove[i], tmpY+yMove[i], tmpDepth+1};
                
                q.offer(newTmp);
                visited[newX][newY] = true;
            }
            
        }
        
        return -1;
        
    }
    
    public int solution(int[][] maps) {
        int answer = 0;
        int n = maps.length;
        int m = maps[0].length;
        visited = new boolean[n][m];
        
        
        return bfs(0,0,n,m,maps);
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    answer = 0
    n = len(maps)
    m = len(maps[0])
    x_move = [1,0,-1,0]
    y_move = [0,1,0,-1]
    visited = [[False for _ in range(m)] for _ in range(n)]
    
    visited[0][0] = True
    q = deque()
    q.append([0,0,0])
    x = 0
    y = 0
    while q:
        current = q.popleft()
        c_x, c_y, c_t = current
        
        if(c_x==n-1 and c_y==m-1): 
            return c_t+1
        
        for i in range(4):
            new_x = c_x + x_move[i]
            new_y = c_y + y_move[i]
            if new_x<0 or new_x>= n or new_y<0 or new_y>=m: continue
            if visited[new_x][new_y] or maps[new_x][new_y]==0: continue
            q.append([new_x, new_y, c_t+1])
            visited[new_x][new_y] = True
    
    return -1